src/Arcus.WebApi.Security/Authentication/SharedAccessKeyAuthenticationAttribute.cs

```diff
@@ -0,0 +1,104 @@
+    ///     Please provide an <see cref="ISecretProvider"/> implementation in the configured services of the request.
+    /// </remarks>
+    public class SharedAccessKeyAuthenticationAttribute : ActionFilterAttribute
+    {
+        private readonly string _headerName, _secretName;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SharedAccessKeyAuthenticationAttribute"/> class.
+        /// </summary>
+        /// <param name="headerName">The name of the request header which value must match the stored secret.</param>
+        /// <param name="secretName">The name of the secret that's being retrieved using the <see cref="ISecretProvider.Get"/> call.</param>
+        public SharedAccessKeyAuthenticationAttribute(string headerName, string secretName)
+        {
+            if (String.IsNullOrWhiteSpace(headerName))
+            {
+                throw new ArgumentException("Header name cannot be blank", nameof(headerName));
+            }
+
+            if (String.IsNullOrWhiteSpace(secretName))
+            {
+                throw new ArgumentException("Secret name cannot be blank", nameof(secretName));
+            }
+
+            _headerName = headerName;
+            _secretName = secretName;
+        }
+        
+        /// <inheritdoc />
+        public override async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context.HttpContext?.Request == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request", 
+                    nameof(context.HttpContext.Request));
+            }
+
+            if (context.HttpContext.Request.Headers == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request headers",
+                    nameof(context.HttpContext.Request.Headers));
+            }
+
+            if (context.HttpContext.RequestServices == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request services", 
+                    nameof(context.HttpContext.RequestServices));
+            }
+
+            var userDefinedSecretProvider = context.HttpContext.RequestServices.GetService<ISecretProvider>();
+            if (userDefinedSecretProvider == null)
+            {
+                throw new KeyNotFoundException(
+                    $"No configured {nameof(ISecretProvider)} implementation found in the request service container. "
+                    + "Please configure such an implementation (ex. in the Startup) of your application");
+            }
+
+            var cachedSecretProvider = new CachedSecretProvider(userDefinedSecretProvider);
```

Comment:
We are creating a new cached provider every time so it's not really cached, is it? Can't we just get an instance from ICachedSecretProvider instead?

Reply:
We could but then we limit the users to only use cached providers. Is that something we should force on users?

Reply:
Personally I would do that to stimulate good consumer behavior - What do you think @SamVanhoutte 	?

Reply:
I think we should stimulate them to use cached or not?

Reply:
Ah, yes we could first test if that returned anything and go for the `ISecretProvider` if not.

Reply:
That's done.

Reply:
Looks perfect, thanks!